src/storage/src/manifest/impl_.rs

```diff
@@ -15,43 +15,72 @@
 use std::marker::PhantomData;
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::Arc;
+use std::time::Duration;
 
 use arc_swap::ArcSwap;
 use async_trait::async_trait;
+use common_runtime::{RepeatedTask, TaskFunction};
 use common_telemetry::{debug, logging, warn};
 use object_store::ObjectStore;
-use snafu::ensure;
+use snafu::{ensure, ResultExt};
 use store_api::manifest::action::{self, ProtocolAction, ProtocolVersion};
 use store_api::manifest::*;
 
-use crate::error::{Error, ManifestProtocolForbidWriteSnafu, Result};
+use crate::error::{
+    Error, ManifestProtocolForbidWriteSnafu, Result, StartManifestGcTaskSnafu,
+    StopManifestGcTaskSnafu,
+};
 use crate::manifest::action::RegionCheckpoint;
 use crate::manifest::checkpoint::Checkpointer;
 use crate::manifest::storage::{ManifestObjectStore, ObjectStoreLogIterator};
 
-// TODO(dennis): export this option to table options or storage options.
-const CHECKPOINT_ACTIONS_MARGIN: u64 = 10;
+const CHECKPOINT_ACTIONS_MARGIN: u16 = 10;
+const GC_DURATION_SECS: u64 = 30;
 
 #[derive(Clone, Debug)]
 pub struct ManifestImpl<S: Checkpoint<Error = Error>, M: MetaAction<Error = Error>> {
     inner: Arc<ManifestImplInner<S, M>>,
     checkpointer: Option<Arc<dyn Checkpointer<Checkpoint = S, MetaAction = M>>>,
     last_checkpoint_version: Arc<AtomicU64>,
+    checkpoint_actions_margin: u16,
+    gc_task: Option<Arc<RepeatedTask<Error>>>,
 }
 
-impl<S: Checkpoint<Error = Error>, M: MetaAction<Error = Error>> ManifestImpl<S, M> {
+impl<S: 'static + Checkpoint<Error = Error>, M: 'static + MetaAction<Error = Error>>
+    ManifestImpl<S, M>
+{
     pub fn new(
         manifest_dir: &str,
         object_store: ObjectStore,
+        checkpoint_actions_margin: Option<u16>,
+        gc_duration: Option<Duration>,
         checkpointer: Option<Arc<dyn Checkpointer<Checkpoint = S, MetaAction = M>>>,
     ) -> Self {
+        let inner = Arc::new(ManifestImplInner::new(manifest_dir, object_store));
+        let gc_task = if checkpointer.is_some() {
```

Comment:
Looks like we will create a task for each manifest? Could we use single task to do the gc work for each region?

Reply:
Yep, we don't want to leak the region logic to the upper storage engine. I think it's fine here. The task is lightweight. But if it's an issue in future, we can refactor it. I'll add a `TODO` here.